solana/solana-ibc/programs/solana-ibc/src/lib.rs

```diff
@@ -402,11 +402,26 @@ pub mod solana_ibc {
         let height = store.borrow().chain.head()?.block_height;
         // height just before the data is added to the trie.
         msg!("Current Block height {}", height);
+        let previous_root = store.borrow().provable.hash().clone();
 
         ::ibc::core::entrypoint::dispatch(&mut store, &mut router, message)
             .map_err(error::Error::ContextError)
             .map_err(move |err| error!((&err)))?;
 
+        #[cfg(feature = "witness")]
+        {
+            let root = store.borrow().provable.hash().clone();
+            if previous_root != root {
+                msg!("Writing local consensus state");
+                let slot = Clock::get()?.slot;
+                let timestamp = Clock::get()?.unix_timestamp as u64;
+                let storage = &mut store.borrow_mut().private;
+                storage
+                    .add_local_consensus_state(slot, timestamp, root)
+                    .unwrap();
```

Comment:
Could also just chain eliminating the local variable:

```suggestion
                store.borrow_mut()
                    .private;
                    .add_local_consensus_state(slot, timestamp, root)
                    .unwrap();
```

Matter of taste I suppose.